### PR TITLE
[ONNX] Handle sequence output shape and type inference

### PR DESCRIPTION
**Pull request by neginraoof**
Handle sequence output shape and type inference.

This PR fixes value type of sequence outputs. Prior to this, all model sequence type outputs were unfolded for ONNX models.
This PR also enable shape inference for sequence outputs to represent the dynamic shape of these values.

### CODE REVIEW

File: torch/csrc/jit/serialization/export.cpp
Comment (BowenBao) [inline torch/csrc/jit/serialization/export.cpp:342]: Need to keep the name or some other way to ensure the dim_param is unique. dim_param with same string are expected to have same size at runtime.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:487]: is `graph->outputs()[graph_output_index]` sequence? on the other hand `outputs[outputs_index]` tensor? how do these two work together

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:445]: together with the below comment, this can be renamed to something like `ONNXUpdateTypeFromTensor`. this may be applied to input as well in the future.

**Review by BowenBao — COMMENTED** [review]
Thanks! great start on covering sequence outputs for onnx models.

File: torch/csrc/jit/serialization/export.cpp
Comment (neginraoof) [reply torch/csrc/jit/serialization/export.cpp:342]: I see. What do you suggest for name? Since node name can be a number, previously I'd see examples like "2_0" where "2" was the debug name. So changing this to "2_dim_0" ?

Comment (BowenBao) [reply torch/csrc/jit/serialization/export.cpp:342]: maybe include optype as well, like "Conv2_dim_0". the optype will be from `value->node()`. need to be careful about edge cases where `value` is graph input (not output from other node). 

not sure if each `value` would have other more displayable `name` available at that stage.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:445]: Assign input shape is in script_init. Is this where you're suggesting to mode the code?
I think shape inference would be the correct place for this.

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:445]: Resolved the comment about use of unflatten.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp]: Updated this part of the code.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:497]: Currently `MergeInferredType` does not handle case merging `List[Tensor]` with `Tensor`. In ONNX, different tensor shape inside one Sequence is supported. And we try best effort to find the compatible representation. Perhaps do something similar with the list of `vars` and find a compatible tensor shape?

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp]: We're inferring the type here. Looks like the output shape for sequences is already handled in export.cpp.

**Comment by BowenBao** [comment]
CI failure is related.
There is another interesting failure related to memory, not sure if it is sporadic or potential memory leak?

```
Nov 04 17:19:02 =================================== FAILURES ===================================
Nov 04 17:19:02 ________________________ TestONNXRuntime.test_groupnorm ________________________
Nov 04 17:19:02 
Nov 04 17:19:02 self = <test_pytorch_onnx_onnxruntime.TestONNXRuntime testMethod=test_groupnorm>
Nov 04 17:19:02 
...
Nov 04 17:19:02 .0 = <zip object at 0x7ff92e891548>
Nov 04 17:19:02 
Nov 04 17:19:02 >   [np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol) for out, ort_out in zip(outputs, ort_outs)]
Nov 04 17:19:02 E   MemoryError: Unable to allocate 534. MiB for an array with shape (139968000,) and data type float32
Nov 04 17:19:02 
Nov 04 17:19:02 test/onnx/test_pytorch_onnx_onnxruntime.py:80: MemoryError
```

File: torch/csrc/jit/serialization/export.cpp
Comment (BowenBao) [inline torch/csrc/jit/serialization/export.cpp:343]: Will need to update expect file, if #46629 goes in first.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:484]: Compared to tuple below, it seems now we are completely supporting list as output? Or in other words, there won't be a case where pt produces list, while onnx produces flattened tensors?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:496]:
```suggestion
        graph->outputs()[outputs_index]->setType(ListType::create(TensorType::create(var)));
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:529]: should we use `ONNXUpdateTypeFromTensor`?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:496]: i'm still unsure here. Can we change this PR to only update scalarType for existing graph output?

TL;DR

Current code will be overwriting graph output type/shape with the one inferred from the first tensor in the list. I think this may not be accurate. 

The most accurate solution would be to go over all the `list_elem`, find the common shape (same logic as in ONNX).

e.g.
`a:[3, 4], b:[4, 4] => c:[None, 4]`
`a:[3, 4, 5], b:[4, 4] => c:None`

And at last merge with existing shape in graph (this is also tricky).

On the other hand i think it's ok not to block this PR with the issue above since several shape inference infras are not ready and are being improved at the moment. We will need to revisit this as we enable more and more sequence use cases.

**Review by BowenBao — COMMENTED** [review]
Looks good, several comments left.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:529]: As we will enable shape inference by default, do you think it makes sense to use ONNXUpdateTypeFromTensor or directly merge with MergeInferredType?
Is there a use-case to keep the conditional?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:514]: We may be able write this function as a recursive call, if tuples or lists could be nested. Let's see if that will be a case.

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:514]: Sure. I haven't encountered an op with forward definition containing a nested vector or tuple of vectors. Though for tuples of vectors or dictionaries of vectors, we can have this in mind to extend this function further if any such cases were seen.
Thank!

**Review by bzinodev — APPROVED** [review]

**Comment by neginraoof** [comment]
@bzinodev This PR is rebased and ready for merge. Thanks!

**Comment by Modnxs** [comment]
O

**Comment by Modnxs** [comment]
#

**Comment by neginraoof** [comment]
@eellison This PR is rebased and ready for merge. Thanks!

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp
Comment (bzinodev) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:497]: should not output be declared as  "const at::Tensor& output" we don't want to copy it every time and it is not changed

Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp]: Makes sense, updated. Thanks.

**Comment by neginraoof** [comment]
@bzinodev Would you please help merge this PR?
Thanks!

**Comment by neginraoof** [comment]
@bzinodev 
Would you please help us merge this PR? This one is the last PR we would like you to merge to master, and it is a high priority PR.

Thanks a lot.

**Review by bzinodev — CHANGES_REQUESTED** [review]
This is a change to ONNX export API as know it can return lists. 

We need more description of this change and discuss if we are break backward compatibility

**Review by bzinodev — APPROVED** [review]

**Comment by neginraoof** [comment]
@bzinodev 
Is it okay to merge this PR now?

**Comment by neginraoof** [comment]
PR to fix broken mypy tests: https://github.com/pytorch/pytorch/pull/50571

**Comment by nairbv** [comment]
Hi @neginraoof, there's an "unland" for this diff that's already in progress (we actively revert changes when tests fail). Can you re-submit the original diff including the fix from 50571 (once the reverting change finishes landing)?

**Pull request closed without merging**
